### PR TITLE
attempt to fix UART0 mikroBUS pin swap

### LINKED ISSUE #67: HW (rev C5): UART pins to mikroBUS headers have swapped TX/RX
According to the [mikroBUS specification](https://download.mikroe.com/documents/standards/mikrobus/mikrobus-standard-specification-v200.pdf), the socket labels are based on the perspective of the MCU, meaning the RX pin on the socket in the following diagram should be connected to the RX pin on the MCU.

image

This is described in more detail in the section below showing how add-on boards should be labelled and connected in the opposite way.

image

This could be swapped in software, except the bootloader is configured in the alternate manner, as well as the connection to the MSP430.

Hey @vaishnav98, wouldn't you have seen this issue on the GPS/GNSS Click boards?

cc @ChrisGammell @statropy @Pillar1989 @FionaYu20180326

### PR DESCRIPTION
**Pull request by vaishnavachath**
Attempt to fix https://github.com/jadonk/beagleconnect/issues/67

Using the emulated I2C slave we control the MB1_CTRL for UART
SPDT Analog Switch.Two endpoints are added to set the UART switch
state and also to get the current state through the I2C slave

Corresponding Zephyr PR : https://github.com/jadonk/zephyr/pull/14

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -346,6 +354,8 @@ int main(void)
         switch (USB_getConnectionState())
         {
             case ST_ENUM_ACTIVE:
+                if(uartmuxState != UARTMUXSEL_MODE_MSP430)
```
Comment (jadonk) [inline main.c]: OK, so the CC1352 locks the MSP430 out? Are we confident this won't happen in common usage scenarios? I think we need a scenario document.

File: main.c [hunk 2]
```diff
@@ -540,15 +550,29 @@ void USCI_B1_ISR (void)
...
+                break;                 
             default:
                 break;
             }
             break;
         }
         case USCI_I2C_UCTXIFG:
             //Transmit data
-            USCI_B_I2C_slavePutData(USCI_B1_BASE, 0xff);
+            bytesReceived = USCI_B_I2C_slaveGetData(USCI_B1_BASE);
```
Comment (jadonk) [inline main.c:609]: What is this change for?

File: main.c [hunk 3]
```diff
@@ -540,15 +550,29 @@ void USCI_B1_ISR (void)
...
+            if(bytesReceived == UARTMUXSEL_GET)
+                USCI_B_I2C_slavePutData(USCI_B1_BASE, uartmuxState);
```
Comment (jadonk) [inline main.c]: This is a bit awkward to be the only parameter that can be read. Probably makes sense to make this more general.

**Comment by jadonk** [comment]
For https://github.com/jadonk/msp430F55xx_usb_uart_bridge/pull/7/files#R391 when the USB cable is disconnected, this is a mux change that is asynchronous and doesn't set the state variable.

File: main.c [hunk 4]
```diff
@@ -85,9 +85,15 @@
 #define UARTMUXSEL_SET_MIKROBUS 0x08
 #define UARTMUXSEL_SET_MSP430 0x09
 #define UARTMUXSEL_GET 0x0C
+#define BCF_I2CSLAVE_GET 0x0D
```
Comment (jadonk) [inline main.c:88]: Please consider dropping the term 'slave' and replacing with 'peripheral' or 'target'.

File: main.c [hunk 5]
```diff
@@ -354,9 +363,26 @@ int main(void)
         switch (USB_getConnectionState())
         {
             case ST_ENUM_ACTIVE:
-                if(uartmuxState != UARTMUXSEL_MODE_MSP430)
+                uartmuxState |= UARTMUXSEL_MODE_LOCK_MSP430;
+                if((uartmuxState & 0x0F) != UARTMUXSEL_MODE_MSP430) {
```
Comment (jadonk) [inline main.c:367]: Why do we need the bitmask? Why not use an enum? Should it really ever have any other bits set?

File: main.c [hunk 6]
```diff
@@ -354,9 +363,26 @@ int main(void)
...
+                     * request UART mux switch by informing CC1352.
+                     * CC1352 is configured to use the BOOT pin as 
+                     * interrupt for the I2C slave, and will read 
+                     * the I2C slave status and further queries
+                     * uartmuxState.
+                     */
+                    i2cSlaveState =  BCF_I2CSLAVE_UARTMUX_CHANGE;                    
+                    if(uartmux_lock_trigger) {
+                        hal_ext_boot(TRUE);
+                        __delay_cycles(2000);
```
Comment (jadonk) [inline main.c:378]: Is this a magic number? What duration is required for the CC1352 to recognize the interrupt?

**Review by jadonk — COMMENTED** [review]
Will this provide previous functionality if we do not update the CC1352 firmware? What is the objective of the LOCK_MSP430 mode and is there any way the CC1352 can lock out the MSP430 in the case it wants to continue talking to a mikroBUS device? Can you make it clear why the CC1352 should have input (so that it can coordinate swapping its TX/RX pins) and not just check the status?

**Comment by vaishnavachath** [comment]
> Will this provide previous functionality if we do not update the CC1352 firmware? What is the objective of the LOCK_MSP430 mode and is there any way the CC1352 can lock out the MSP430 in the case it wants to continue talking to a mikroBUS device? Can you make it clear why the CC1352 should have input (so that it can coordinate swapping its TX/RX pins) and not just check the status?

- Yes, if CC1352 firmware is not updated, then MSP430 will only trigger a I2C read request by toggling the boot line, so if CC1352 FW is not updated then no action will be taken

- LOCK_MSP430 -> this is a state to indicate the condition where USB is active, so the UART mux state will be locked to enable the USB-UART commnication

- when LOCK_MSP430 is active then even if CC1352 requests to change the UART mux state to communicate to mikroBUS device, this request won't be handled i.e whenever USB is active , USB-UART state is selected.

- CC1352 initiating the UART mux transition enables us to have this control in the application, e.g greybus application can request for the UART switch in the application itself.. MSP430 taking the decision and triggering the swap according to USB state is also a feasible approach(only difference I can think of is that in this case if CC1352 misses to service the interrupt, then it can end up in a TX-TX connected state and during the transition also this can happen, but in the current implementation this will not happen. 
-

File: main.c (hunk 1 above)
Comment (vaishnavachath) [reply main.c]: Momentarily this happens when current state is mikrobus and USB is connected, now MSP430 will trigger I2C read request by toggling CC1352 boot line,  CC1352 will immediately read the status and swap the TX/RX lines and the UART mux state will be changed to USB-UART

File: main.c (hunk 2 above)
Comment (vaishnavachath) [reply main.c:609]: This enables I2C register reads from the emulated I2C peripheral on MSP430

File: main.c (hunk 3 above)
Comment (vaishnavachath) [reply main.c]: This is fixed now in the latest commit

File: main.c (hunk 4 above)
Comment (vaishnavachath) [reply main.c:88]: Sure, will fix that

File: main.c (hunk 6 above)
Comment (vaishnavachath) [reply main.c:378]: Will update in the next commit, reused the same values used for the bootloader trigger delay, delay to recognize the interrupt should be less than this.

File: main.c (hunk 1 above)
Comment (jadonk) [reply main.c]: OK, so, with CC1352 currently applying the UART pin swap fix, the order is:

1. USB connection detected by MSP430
2. MSP430 sends BOOT interrupt to CC1352
3. CC1352 reads USB connection status
4. CC1352 disables the RX/TX swap
5. CC1352 sends message to MSP430 to flip the mux, giving MSP430 access to the UART

To protect the mikroBUS add-on board and CC1352, should the CC1352 perform the swap by

1. Disable TX and swap RX
2. Send the message to the MSP430
3. Enable TX only after getting traffic on the RX line

?

File: main.c [hunk 7]
```diff
@@ -82,6 +82,13 @@
 #define  LED3_ON 0x06
 #define  LED3_OFF 0x07
 
+#define UARTMUXSEL_SET_MIKROBUS 0x08
+#define UARTMUXSEL_SET_MSP430 0x09
+#define UARTMUXSEL_GET 0x0C
+
+#define UARTMUXSEL_MODE_MIKROBUS 0x0A
+#define UARTMUXSEL_MODE_MSP430   0x0B
+
```
Comment (jadonk) [inline main.c:97]: It is slightly confusing that the commands are different than the mode bits and that the mode bits aren't simply 0/1, but I think this is OK.

File: main.c [hunk 8]
```diff
@@ -540,15 +550,29 @@ void USCI_B1_ISR (void)
...
             break;
         }
         case USCI_I2C_UCTXIFG:
             //Transmit data
-            USCI_B_I2C_slavePutData(USCI_B1_BASE, 0xff);
+            bytesReceived = USCI_B_I2C_slaveGetData(USCI_B1_BASE);
+            if(bytesReceived == UARTMUXSEL_GET)
+                USCI_B_I2C_slavePutData(USCI_B1_BASE, uartmuxState);
+            else
+                USCI_B_I2C_slavePutData(USCI_B1_BASE, 0xff);
```
Comment (jadonk) [inline main.c]: Does this mean that all I2C commands now provide data back?

File: main.c [hunk 9]
```diff
@@ -540,15 +550,29 @@ void USCI_B1_ISR (void)
...
+                uartmuxState = UARTMUXSEL_MODE_MSP430;
+                hal_ext_uart(true);
+                hal_ext_led(3, 1);
+                break;                 
             default:
                 break;
             }
             break;
         }
         case USCI_I2C_UCTXIFG:
```
Comment (jadonk) [inline main.c:607]: Is this state within `switch (USB_getConnectionState())`? Why would USB_getConnectionState() provide any state relative to I2C? Shouldn't this just be another test within the while(1) loop?

**Review by jadonk — COMMENTED** [review]
It seems I missed where the USB function returned something related to the I2C state. I'm going to look for that. Once I can understand what is going on here, I'll merge.

**Comment by jadonk** [comment]
OK, now that I've read it at https://github.com/vaishnav98/msp430F55xx_usb_uart_bridge/blob/mbuart-pinswap/main.c, it makes sense to me. The way it read here made it look like the case statement was part of the while(1) in main and not part of the I2C ISR. I'll merge and we can do any clean-up post-merge.

**Pull request merged**